### ISSUE TITLE
[RFE] Improve CAPR error messaging -- vague certificate errors encountered do not provide enough context to effectively debug CAPR

### ISSUE THREAD

**Issue by flhelfer** [rancher/rancher#41650]
This issue will serve to track an improvement in error messaging when CAPR/K3s/RKE2 are not functioning properly. The most noisy error message encountered in the logs of the `rancher-system-agent` will generally state that there are certificates missing for `kube-scheduler` and `kube-controller-manager`; however, this is expected as CAPR relies on the corresponding Kubernetes component to generate a self-signed certificate.

These certificates are used by CAPR to determine the health of the corresponding components. Without the certificates, we cannot reliably connect to the `healthz` endpoint of the component.

Because of how bootstrapping of K8s components happens within RKE2, if there is a problem starting the container runtime or the components are unable to start, the user is left with their `rancher-system-agent` continuously reporting that the certificates were not generated.

In this case, the correct course of action is to
1. Verify RKE2 is properly installed
2. Verify RKE2 has started and the components have started

Misconfiguration of RKE2, bad networking configurations, overly restrictive systems (read-only file systems, selinux policies, etc.) can all lead to RKE2 being unable to bootstrap the components into place, thus leading to the certificates never being generated.

As such, I think the correct thing to do here would be add additional logic to CAPR + system-agent to check the status of RKE2 and report the problems as such. There realistically is no reason to run the probes against the components until RKE2 has started/is running. 

For users that are encountering this issue, if you believe your issue to be unique, please collect as much data as possible from your machines. A logs collector dump is likely the most helpful: https://github.com/rancherlabs/support-tools/tree/master/collection/rancher/v2.x/logs-collector

### Original Issue Contents Below: 

> 
> **Rancher Server Setup**
> - Rancher version: 2.7.3
> - Installation option (Docker install/Helm Chart): HELM
>    - K3S v1.25.9+k3s1
>    - 
> **Information about the Cluster**
> - Kubernetes version: v1.25.9+rke2r1
> - Cluster Type (Downstream):
>    - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider):
>    -  Custom
>    - Node VM OS Ubuntu 22.04
> 
> **User Information**
> - What is the role of the user logged in? Admin
> 
> **Describe the bug**
> When creating a new custom cluster via Rancher's UI, the nodes that I try to add are stuck in "reconcilling" sate.
> 
> **To Reproduce**
> 1. Create a new cluster on Rancher UI following documentation:
> https://ranchermanager.docs.rancher.com/reference-guides/cluster-configuration/rancher-server-configuration/rke2-cluster-configuration
> All parameters to default except:
> Kubernetes Version: v1.25.9+rke2r1
>  Container Network: Calico
> NGINX Ingress: Selected or not I had the same problem
> 
> 4. Register a node
> Keep all roles selected in step 1
> Select "Insecure" in step 2
> Copy and paste the cmd to the first node
> 
> **Result**
> The node stay in "reconcilling" state with this message: "Waiting for probes: kube-apiserver, kube-controller-manager, kube-scheduler"
> 
> Command to obtain logs on the node:
> `sudo journalctl -u  rancher-system-agent.service -f`
> 
> Logs obtained:
> `May 24 13:26:56 s001kubed01 rancher-system-agent[16705]: time="2023-05-24T13:26:56Z" level=error msg="error loading CA cert for probe (kube-scheduler) /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt: open /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt: no such file or directory"
> May 24 13:26:56 s001kubed01 rancher-system-agent[16705]: time="2023-05-24T13:26:56Z" level=error msg="error while appending ca cert to pool for probe kube-scheduler"
> May 24 13:26:57 s001kubed01 rancher-system-agent[16705]: time="2023-05-24T13:26:57Z" level=error msg="error encountered during parsing of last run time: parsing time \"\" as \"Mon Jan _2 15:04:05 MST 2006\": cannot parse \"\" as \"Mon\""
> May 24 13:26:57 s001kubed01 rancher-system-agent[16705]: time="2023-05-24T13:26:57Z" level=error msg="error loading CA cert for probe (kube-controller-manager) /var/lib/rancher/rke2/server/tls/kube-controller-manager/kube-controller-manager.crt: open /var/lib/rancher/rke2/server/tls/kube-controller-manager/kube-controller-manager.crt: no such file or directory"
> May 24 13:26:57 s001kubed01 rancher-system-agent[16705]: time="2023-05-24T13:26:57Z" level=error msg="error while appending ca cert to pool for probe kube-controller-manager"
> May 24 13:26:57 s001kubed01 rancher-system-agent[16705]: time="2023-05-24T13:26:57Z" level=error msg="error loading CA cert for probe (kube-scheduler) /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt: open /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt: no such file or directory"
> May 24 13:26:57 s001kubed01 rancher-system-agent[16705]: time="2023-05-24T13:26:57Z" level=error msg="error while appending ca cert to pool for probe kube-scheduler"
> May 24 13:27:01 s001kubed01 rancher-system-agent[16705]: time="2023-05-24T13:27:01Z" level=error msg="error encountered during parsing of last run time: parsing time \"\" as \"Mon Jan _2 15:04:05 MST 2006\": cannot parse \"\" as \"Mon\""
> `
> 
> Directories "/var/lib/rancher/rke2/server/tls/kube-controller-manager" and "/var/lib/rancher/rke2/server/tls/kube-scheduler" are indeed missing. So are the CRTs (not even misplaced)
> 
> **Expected Result**
> The node get registered automatically
> 
> Thanks for your help 😄

**Comment by SorsOps** [comment]
I get this error, but I catted the file, it does indeed exist . So as @Oats87  mentioned it was transient 

Checking the kubelet log I see that the Calico overlay is never starting. I tried everything I could to try work out why this was happening to no avail. Eventually I read a[ stack overflow post ](https://stackoverflow.com/questions/63367988/kubernetes-worker-node-is-notready-due-to-cni-plugin-not-initialized#63368818) which mentioned that it is likely due to resource starvation. I updated my memory on the node to 4 gigs from 2 and it started working

I used the same config as you  @flhelfer

**Comment by benz9527** [comment]
```bash
Jul 07 12:07:43 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:43Z" level=info msg="[Applyinator] Running command: sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null]"
Jul 07 12:07:43 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:43Z" level=info msg="[Applyinator] Command sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null] finished with err: <nil> and exit code: 127"
Jul 07 12:07:44 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:44Z" level=error msg="error while appending ca cert to pool for probe kube-scheduler"
Jul 07 12:07:44 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:44Z" level=error msg="error while appending ca cert to pool for probe kube-controller-manager"
Jul 07 12:07:44 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:44Z" level=error msg="error loading x509 client cert/key for probe kube-apiserver (/var/lib/rancher/rke2/server/tls/client-kube-apiserver.crt//var/lib/rancher/rke2/server/tls/client-kube-apiserver.key): tls: failed to find any PEM data in certificate input"
Jul 07 12:07:44 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:44Z" level=error msg="error while appending ca cert to pool for probe kube-apiserver"
Jul 07 12:07:44 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:44Z" level=info msg="[K8s] updated plan secret fleet-default/custom-c7ffee61e8c4-machine-plan with feedback"
Jul 07 12:07:44 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:44Z" level=error msg="error encountered during parsing of last run time: parsing time \"\" as \"Mon Jan _2 15:04:05 MST 2006\": cannot parse \"\" as \"Mon\""
Jul 07 12:07:44 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:44Z" level=error msg="error while appending ca cert to pool for probe kube-scheduler"
Jul 07 12:07:44 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:44Z" level=error msg="error while appending ca cert to pool for probe kube-controller-manager"
Jul 07 12:07:44 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:44Z" level=error msg="error loading x509 client cert/key for probe kube-apiserver (/var/lib/rancher/rke2/server/tls/client-kube-apiserver.crt//var/lib/rancher/rke2/server/tls/client-kube-apiserver.key): tls: failed to find any PEM data in certificate input"
Jul 07 12:07:44 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:44Z" level=error msg="error while appending ca cert to pool for probe kube-apiserver"
Jul 07 12:07:49 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:49Z" level=error msg="error encountered during parsing of last run time: parsing time \"\" as \"Mon Jan _2 15:04:05 MST 2006\": cannot parse \"\" as \"Mon\""
Jul 07 12:07:49 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:49Z" level=error msg="error loading x509 client cert/key for probe kube-apiserver (/var/lib/rancher/rke2/server/tls/client-kube-apiserver.crt//var/lib/rancher/rke2/server/tls/client-kube-apiserver.key): tls: failed to find any PEM data in certificate input"
Jul 07 12:07:49 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:49Z" level=error msg="error while appending ca cert to pool for probe kube-controller-manager"
Jul 07 12:07:49 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:49Z" level=error msg="error while appending ca cert to pool for probe kube-apiserver"
Jul 07 12:07:49 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:49Z" level=error msg="error while appending ca cert to pool for probe kube-scheduler"
Jul 07 12:07:54 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:54Z" level=error msg="error encountered during parsing of last run time: parsing time \"\" as \"Mon Jan _2 15:04:05 MST 2006\": cannot parse \"\" as \"Mon\""
Jul 07 12:07:54 k8s13ms1 rancher-system-agent[23909]: time="2023-07-07T12:07:54Z" level=error msg="error loading x509 client cert/key for probe kube-apiserver (/var/lib/rancher/rke2/server/tls/client-kube-apiserver.crt//var/lib/rancher/rke2/server/tls/client-kube-apiserver.key): tls: failed to find any PEM data in certificate input"
```
rancher v2.7.5
kubernetes v1.26.6+rke2r1

**Comment by Humungus14** [comment]
Got same issue on rancher v2.6.13.
k8s versions: v1.22.17+rke2r1, v1.23.17+rke2r1

**Comment by eciarm** [comment]
Got same issue on 
rancher v2.6.10
k8s versions: v1.22.17+rke2r1, v1.23.15+rke2r1, v1.24.9+rke2r2 
ubuntu 20,04.6 lts

**Comment by cboergermann** [comment]
Same here:

Rancher 2.7.5
v1.24.15+rke2r1 
ubuntu 22.04.2

**Comment by cristi-vlad** [comment]
Same problem here

Rancher 2.7.5
v1.26.5+rke2r1
redhat 8.7 
docker client: 23.0.1
docker server: 19.03.14

**Comment by GuillaumeOuint** [comment]
Same problem here

Rancher 2.7.5
v1.26.6+rke2r1
Ubuntu 22.04.2

**Comment by GuillaumeOuint** [comment]
Actually, I think I found the problem, when we provision a dual stack cluster with the rancher UI, we can't use dual stack pod/service cidr because we can't register the node with multiple IP (bogus for tcp listen because the text is like "fc00::1,192.168.0.1"). And when I try to register the node with only one IP, I get an error telling me that the node ip has to be of the same family of the pod cidr (ipv4 + ipv6). Doesn't make any sense.

**Comment by enterdv** [comment]
Same issue

Rancher 2.7.5
v1.26.6+rke2r1
Ubuntu 22.04

**Comment by enterdv** [comment]
I fixed the issue with disabling IPv6 IP-addresses on master nodes. When the first node was bootstrapped it tried to connect with two other node with IPv6 and failed. After disabling all works fine.

**Comment by benz9527** [comment]
> I fixed the issue with disabling IPv6 IP-addresses on master nodes. When the first node was bootstrapped it tried to connect with two other node with IPv6 and failed. After disabling all works fine.

In my env, I need the dual-stack ip addresses.

**Comment by GuillaumeOuint** [comment]
> > I fixed the issue with disabling IPv6 IP-addresses on master nodes. When the first node was bootstrapped it tried to connect with two other node with IPv6 and failed. After disabling all works fine.
> 
> In my env, I need the dual-stack ip addresses.

You can check [my repo](https://github.com/GuillaumeOuint/rancher) where I've fixed the issue (WORKING DUAL STACK !!), if you intend to use this code as is, you have to change the rancher agent version to v2.7-head on line 40 (var VERSION) of the file cmd/agent/main.go

**Comment by GuillaumeOuint** [comment]
> > > I fixed the issue with disabling IPv6 IP-addresses on master nodes. When the first node was bootstrapped it tried to connect with two other node with IPv6 and failed. After disabling all works fine.
> > 
> > 
> > In my env, I need the dual-stack ip addresses.
> 
> You can check [my repo](https://github.com/GuillaumeOuint/rancher) where I've fixed the issue (WORKING DUAL STACK !!), if you intend to use this code as is, you have to change the rancher agent version to v2.7-head on line 40 (var VERSION) of the file cmd/agent/main.go

And you may also want to use the kubelet arg --node-ip=:: to make kubelet advertise his IPv6 in k8s Node object.

**Comment by GuillaumeOuint** [comment]
I've created PR https://github.com/rancher/rancher/pull/42422 for fixing dual stack cluster creation

**Comment by benz9527** [comment]
> > > I fixed the issue with disabling IPv6 IP-addresses on master nodes. When the first node was bootstrapped it tried to connect with two other node with IPv6 and failed. After disabling all works fine.
> > 
> > 
> > In my env, I need the dual-stack ip addresses.
> 
> You can check [my repo](https://github.com/GuillaumeOuint/rancher) where I've fixed the issue (WORKING DUAL STACK !!), if you intend to use this code as is, you have to change the rancher agent version to v2.7-head on line 40 (var VERSION) of the file cmd/agent/main.go

TKS

**Comment by qRainBirdp** [comment]
Same problem here

Rancher 2.7.6
v1.26.7+rke2r1
Centos7.8
I tried disabling the IPv6 of the node, but I still couldn't register the node.

**Comment by Riadh-Tlidjine** [comment]
Same problem here

Rancher 2.7.é
v1.25.7+rke2r1
rh8.6
I tried disabling the IPv6 of the node, but I still couldn't register the node.
Any idea ?

**Comment by GuillaumeOuint** [comment]
@Riadh-Tlidjine @qRainBirdp A PR is open about this (very easy to integrate given the code changes) since a month, but the developers don't care much about it because they haven't responded yet. Even though this problem is critical and shouldn't be in first place because IPv6 is almost my age. Shameful

**Comment by dmpe** [comment]
Can also confirm, on 2.7.6 (centos 7) and RKE2 downstream cluster 1.24. IPv6 is already disabled on my VMs and still can observe the issue

**Comment by Riadh-Tlidjine** [comment]
Hi, good news, we have find the root cause,

We had the options "noexec" on /var mountpoint in /etc/fstab, so some execution scripts could'nt be executed.

/dev/mapper/rootvg-var  /var                    ext4    defaults,noexec,nosuid,nodev 1 2

This is because we are using already hardened RHEL iso version.

**Comment by martin-fiedler** [comment]
The Problem still occurs and it is not a fstab error

**Comment by gyohng** [comment]
Same problem.

**Comment by jbertozzi** [comment]
Same issue, mostly default parameters used.

**Comment by gapopp** [comment]
same issue on Rancher 2.7.9, Calico, 1.26.8+rke2r1

**Comment by Hsiao-Chien-Ti** [comment]
Same here.

**Comment by arusa** [comment]
> It's interesting that this problem occurs (at least in my case) only when I'm using RockyLinux 9.2... I tried the exact same stack but this time Ubuntu 20.04.6LTS and everything went as it should.

With Ubuntu 22.04 it doesn't work for me.

**Comment by rajivreddy** [comment]
Had the same issue with ubuntu 22.04
```logs
Feb 02 11:47:58 ip-172-31-0-143 rancher-system-agent[17366]: time="2024-02-02T11:47:58Z" level=error msg="error while appending ca cert to pool for probe kube-controller-manager"
Feb 02 11:47:58 ip-172-31-0-143 rancher-system-agent[17366]: time="2024-02-02T11:47:58Z" level=error msg="error loading CA cert for probe (kube-scheduler) /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt: open /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt: no such file or directory"
```

**Comment by dcardozoo** [comment]
For us, we had to manually created the etcd user and additionally make sure to set the kernel parameters https://docs.k3s.io/security/hardening-guide#set-kernel-parameters 

Kubernetes 1.24.17 - cis 1.6

**Comment by RaijinLin** [comment]
if your cluster is ready ,just use Import Existing -> Generic (or somethink you want) -> Registration

**Comment by timothystewart6** [comment]
This also happens to me with 
Rancher v2.8.2, RKE2 v1.27.12+rke2r1, calico
Downstream Cluster: RKE2, v1.27.12+rke2r1, , calico

Registered 3 etcd / control nodes, 3 workers

First node partially registers but then gets into an error loop of 

>  dial tcp 127.0.0.1:6443: connect: connection refused

and cannot proceed.

this has happened on multiple downstream clusters (RKE2 and K3s) and happens no matter which installed components or roles I choose.

The only fix I have found is to wait long enough and reboot the node trying to register.  Eventually it will delete old pods, delete old containers, and try again.

TLDR, if you see this in the logs when running

`tail -f /var/lib/rancher/rke2/agent/logs/kubelet.log` just keep rebooting until it properly registers.

**Comment by alphabet5** [comment]
This is still an issue with k8s 1.28.8-rke2r1 and Rancher 2.8.3

This still fixes it for me.

```
timeout 1 openssl s_client -connect 127.0.0.1:10259 -showcerts 2>&1 | grep -A 19 -m 1 'BEGIN CERTIFICATE' | sudo tee /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt
timeout 1 openssl s_client -connect 127.0.0.1:10257 -showcerts 2>&1 | grep -A 19 -m 1 'BEGIN CERTIFICATE' | sudo tee /var/lib/rancher/rke2/server/tls/kube-controller-manager/kube-controller-manager.crt
```

**Comment by kgtw** [comment]
Also seeing the same issues using Rancher `v2.8.3` during an upgrade from RKE2 `v1.27.6+rke2r1` to `v1.27.13+rke2r1`

**Comment by kgtw** [comment]
To add some additional context, we managed to upgrade each patch version from `v1.27.6+rke2r1` to `v1.27.12+rke2r1` without any issues. The problem for us is introduced with the [change set](https://github.com/rancher/rke2/compare/v1.27.12+rke2r1...v1.27.13+rke2r1) between `v1.27.12+rke2r1` and `v1.27.13+rke2r1`.

**Comment by Edcursion-DavidN** [comment]
We are also facing this same issue.

Racher Version: v2.8.4
- Hosted via Docker on Ubuntu Server 24.04 LTS

Cluster Config:
- Tested Kubernetes Version: v1.28.9+rke2r1 and v1.26.15+rke2r1
- Tested Container Network: Calico and Cilium

All Nodes and the rancher server are hosted on Hyper-V using Ubuntu Server LTS 24.04 within the same V Switch.

**Comment by asaura08** [comment]
For me it was VPS provider firewall

**Comment by riuvshyn** [comment]
@asaura08 
hmmm maybe then it was net policies for us too...
cc @kgtw

**Comment by zlingqu** [comment]
Had the same issue with ubuntu 22.04. 
If you started rancher server with docker run, please note the version compatibility of rancher and docker, see here: [https://www.suse.com/suse-rancher/support-matrix/all-supported-versions/rancher-v2-8-5/](https://www.suse.com/suse-rancher/support-matrix/all-supported-versions/rancher-v2-8-5/),
For example, rancher 2.8.5, when I use docker 27.1.1, it's not OK, but when I switch to docker 24.0.9 , it's OK.

**Comment by olekgo** [comment]
In my case, the issue was likely related to networking. Although I haven't fully researched it yet, it seems that the server I tried to register could reach Rancher but couldn't access the internet (both are in the same subnet), possibly due to a misconfigured default gateway

**Comment by Linsong666** [comment]
> In my case, the issue was likely related to networking. Although I haven't fully researched it yet, it seems that the server I tried to register could reach Rancher (the server and rancher are in the same subnet) but couldn't reach the internet, likely due to a misconfigured default gateway.

How did you solve this problem?

**Comment by olekgo** [comment]
> > In my case, the issue was likely related to networking. Although I haven't fully researched it yet, it seems that the server I tried to register could reach Rancher (the server and rancher are in the same subnet) but couldn't reach the internet, likely due to a misconfigured default gateway.
> 
> How did you solve this problem?

I fixed the internet connectivity by fixing the default route (I had two, I had to remove the incorrect one using `ip route del default via 10.42.0.1` command). This is very specific to my setup, however, the gist is that if you get this issue - check you internet, ensure your machine can download k3s/rke.

**Comment by kgtw** [comment]
It might not be related to the exact same problem that OP had, but our problem was resolved after removing the extra host-path mounts for control-plane components. Something changed between `v1.27.12+rke2r1` and `v1.27.13+rke2r1` where automatic discovery was happening for host-paths based on the components args/configuration inside the containers, this was resulting in multiple mounts within the pod spec for the same paths.

Removing the following configuration from RKE2 configuration was the fix:
- `kube-apiserver-extra-mount`
- `kube-scheduler-extra-mount`
- `kube-controller-manager-extra-mount`
- `kube-proxy-extra-mount`
- `etcd-extra-mount`

**Comment by NoVeTe36** [comment]
> Had the same issue with ubuntu 22.04. If you started rancher server with docker run, please note the version compatibility of rancher and docker, see here: https://www.suse.com/suse-rancher/support-matrix/all-supported-versions/rancher-v2-8-5/, For example, rancher 2.8.5, when I use docker 27.1.1, it's not OK, but when I switch to docker 24.0.9 , it's OK.

Many thanks. I followed your suggestion and it worked (docker 26.0.1, Ubuntu 24.04)

**Comment by zlingqu** [comment]
- I met there is a scene and the type of situation, I use the docker run to start the rancher, started using the parameters `- e CATTLE_SYSTEM_DEFAULT_REGISTRY=harbor.xxx.com: 55001/proxy_dockerhub`.

- This parameter ensures that my rancher server pulls images from my private repository, which is built with goharbor. When I was at the rancher UI new cluster, there is a registries parameters, will be the default fill `harbor.xxx.com: 55001/proxy_dockerhub`, I use the default.

- When a node is added to a cluster, My `journalctl -xu rancher-system-agent -f` repeatedly error ` Certs and the directory/var/lib/rancher/rke2 / server/TLS/kube - the scheduler and The/var/lib/rancher/rke2 / server/TLS/kube - controller - manager missing `, then I found the`/var/lib/rancher/rke2/server/TLS` not generated.

- At the same time, I noticed that rancher-system-agent would start the rke2-server service, so I looked at `journalctl -xu rke2-server-f`, I see the error `level=fatal msg="failed to parse system-default-registry: registries must be valid RFC 3986 URI authorities: Harbor.xxx.com: 55001/proxy_dockerhub`, rk2 server did not rise.

- At this point I realized that there was a registry format problem, that the url should not have a path behind it, and when I used docker registry to rebuild a proxy and support https, everything went back to normal. Of course, registries also supports other advanced parameters when adding cluster, which can also solve this problem, but I didn't realize it at the time.

- I wonder if we should perform a matching and interception on rancher ui to ensure that registry meets the format requirements, or check and report errors in Rancher-system-agent

- [ ] docker: 24.0.9
- [ ] rancher: 2.8.5
- [ ] rke2: v1.28.11+rke2r1

**Comment by NoVeTe36** [comment]
Have to go back to this issue =))) When a master node is added on same as the machine that installed rancher, it runs OK but when register another machine to the cluster, it stills stuck at `[INFO ] configuring bootstrap node(s) custom-ff8710372642: waiting for probes: etcd, kube-apiserver, kube-controller-manager, kube-scheduler, kubelet`. 

On the node that i want to add, the **rke2-server** is failed to start with this log:
```
Aug 19 07:58:31 k8s03 systemd[1]: rke2-server.service: Unit process 241451 (containerd-shim) remains running after unit stopped.
Aug 19 07:58:31 k8s03 systemd[1]: Failed to start Rancher Kubernetes Engine v2 (server).
```

And the **rancher-system-agent** just keep repeating **[K8s] updated plan secret fleet-default/custom-[number]-machine-plan with feedback** as follows

```
Aug 19 08:08:51 k8s03 rancher-system-agent[250471]: time="2024-08-19T08:08:51Z" level=info msg="[Applyinator] Command sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null] finished with err: <nil> and exit code: 1"
Aug 19 08:08:56 k8s03 rancher-system-agent[250471]: time="2024-08-19T08:08:56Z" level=info msg="[K8s] updated plan secret fleet-default/custom-ff8710372642-machine-plan with feedback"
Aug 19 08:08:56 k8s03 rancher-system-agent[250471]: time="2024-08-19T08:08:56Z" level=error msg="[K8s] received secret to process that was older than the last secret operated on. (69678 vs 70659)"
Aug 19 08:08:56 k8s03 rancher-system-agent[250471]: time="2024-08-19T08:08:56Z" level=error msg="error syncing 'fleet-default/custom-ff8710372642-machine-plan': handler secret-watch: secret received was too old, requeuing"
Aug 19 08:11:53 k8s03 rancher-system-agent[250471]: time="2024-08-19T08:11:53Z" level=info msg="[Applyinator] No image provided, creating empty working directory /var/lib/rancher/agent/work/20240819-081153/c8de93a2ee2e8bc39534226a16d3e12a629f7ea58c5d00aa2076b26588c4bf52_0"
Aug 19 08:11:53 k8s03 rancher-system-agent[250471]: time="2024-08-19T08:11:53Z" level=info msg="[Applyinator] Running command: sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null]"
Aug 19 08:11:53 k8s03 rancher-system-agent[250471]: time="2024-08-19T08:11:53Z" level=info msg="[Applyinator] Command sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null] finished with err: <nil> and exit code: 1"
Aug 19 08:11:58 k8s03 rancher-system-agent[250471]: time="2024-08-19T08:11:58Z" level=info msg="[K8s] updated plan secret fleet-default/custom-ff8710372642-machine-plan with feedback"
Aug 19 08:11:58 k8s03 rancher-system-agent[250471]: time="2024-08-19T08:11:58Z" level=error msg="[K8s] received secret to process that was older than the last secret operated on. (70659 vs 71636)"
Aug 19 08:11:58 k8s03 rancher-system-agent[250471]: time="2024-08-19T08:11:58Z" level=error msg="error syncing 'fleet-default/custom-ff8710372642-machine-plan': handler secret-watch: secret received was too old, requeuing"
```

**Comment by NoVeTe36** [comment]
idk what just happened but i just simply delete every thing using `which` with `rancher-system-agent`, `rke2-server` and relevant things. After that, delete all that, also delete `/var/lib/rancher` and `systemctl daemon-reload` to reset or simply reboot. 

Afterwards, re-install rancher, i assume, on node 1. The version that I installed is 
```
- docker: 24.0.9
- rancher: stable (v2.8.5)
- kubernets: v1.28.11 +rke2r1
- ubuntu: 22.04
```

Install with the given registration on node 2, through stills can't find ca-cert, however, it runs:
```
Aug 19 14:14:09 k8s03 rancher-system-agent[2673]: time="2024-08-19T14:14:09Z" level=error msg="error while appending ca cert to pool for probe kube-controller-manager"
Aug 19 14:14:09 k8s03 rancher-system-agent[2673]: time="2024-08-19T14:14:09Z" level=error msg="error loading CA cert for probe (kube-scheduler) /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt: open /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt: no such file or directory"
Aug 19 14:14:09 k8s03 rancher-system-agent[2673]: time="2024-08-19T14:14:09Z" level=error msg="error while appending ca cert to pool for probe kube-scheduler"
Aug 19 14:14:09 k8s03 rancher-system-agent[2673]: time="2024-08-19T14:14:09Z" level=info msg="[K8s] updated plan secret fleet-default/custom-ff8710372642-machine-plan with feedback"
Aug 19 14:14:09 k8s03 rancher-system-agent[2673]: time="2024-08-19T14:14:09Z" level=error msg="error loading CA cert for probe (kube-controller-manager) /var/lib/rancher/rke2/server/tls/kube-controller-manager/kube-controller-manager.crt: open /var/lib/rancher/rke2/server/tls/kube-controller-manager/kube-controller-manager.crt: no such file or directory"
Aug 19 14:14:09 k8s03 rancher-system-agent[2673]: time="2024-08-19T14:14:09Z" level=error msg="error while appending ca cert to pool for probe kube-controller-manager"
Aug 19 14:14:09 k8s03 rancher-system-agent[2673]: time="2024-08-19T14:14:09Z" level=error msg="error loading CA cert for probe (kube-scheduler) /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt: open /var/lib/rancher/rke2/server/tls/kube-scheduler/kube-scheduler.crt: no such file or directory"
Aug 19 14:14:09 k8s03 rancher-system-agent[2673]: time="2024-08-19T14:14:09Z" level=error msg="error while appending ca cert to pool for probe kube-scheduler"
Aug 19 14:14:14 k8s03 rancher-system-agent[2673]: time="2024-08-19T14:14:14Z" level=info msg="[K8s] updated plan secret fleet-default/custom-ff8710372642-machine-plan with feedback"
Aug 19 14:16:21 k8s03 rancher-system-agent[2673]: time="2024-08-19T14:16:21Z" level=info msg="[K8s] updated plan secret fleet-default/custom-ff8710372642-machine-plan with feedback"
```

There also some references useful for me: **[41125](https://github.com/rancher/rancher/issues/41125#issuecomment-1506620040)**, **[compatible version](https://github.com/rancher/rancher/issues/41650#issuecomment-2255350740)** and **[certification](https://github.com/rancher/rancher/issues/41650#issuecomment-2075384843)**

Pardon me if i'm wrong :sweat_smile: and my markdown is a little bit limited :stuck_out_tongue:

**Comment by dustin-groh-dev** [comment]
I'm experiencing this with a customer attempting to create a k3s cluster via Rancher.

k3s 1.30.6
Rancher 2.9.2

```
rancher-system-agent[1637075]: time="2024-12-10T14:33:17-05:00" level=error msg="error loading CA cert for probe (kube-scheduler) /var/lib/rancher/k3s/server/tls/kube-scheduler/kube-scheduler.crt: open /var/lib/rancher/k3s/server/tls/kube-scheduler/kube-scheduler.crt: no such file or directory"
```
In fact `/var/lib/rancher/k3s/server/tls/kube-scheduler/kube-scheduler.crt` does not exist in this environment, and all other certs necessary do exist.

Here are some lines from the provisioning log in Rancher
```
[INFO ] waiting for at least one control plane, etcd, and worker node to be registered
[INFO ] waiting for viable init node
[INFO ] configuring bootstrap node(s) custom-9bdc82db3aed: waiting for agent to check in and apply initial plan
[INFO ] configuring bootstrap node(s) custom-9bdc82db3aed: error applying plan -- check rancher-system-agent.service logs on node for more information, waiting for agent to check in and apply initial plan
```

Reading through the thread here it seems there isn't in a consensus on what can cause this problem directly other than "network issues". I was able to recreate the issue in my lab reusing a node from a previous cluster and then using a node with a clean install the node joins successfully, not sure what the common denominator is.

**Comment by NoVeTe36** [comment]
> I'm experiencing this with a customer attempting to create a k3s cluster via Rancher.
> 
> k3s 1.30.6 Rancher 2.9.2
> 
> ```
> rancher-system-agent[1637075]: time="2024-12-10T14:33:17-05:00" level=error msg="error loading CA cert for probe (kube-scheduler) /var/lib/rancher/k3s/server/tls/kube-scheduler/kube-scheduler.crt: open /var/lib/rancher/k3s/server/tls/kube-scheduler/kube-scheduler.crt: no such file or directory"
> ```
> 
> In fact `/var/lib/rancher/k3s/server/tls/kube-scheduler/kube-scheduler.crt` does not exist in this environment, and all other certs necessary do exist.
> 
> Here are some lines from the provisioning log in Rancher
> 
> ```
> [INFO ] waiting for at least one control plane, etcd, and worker node to be registered
> [INFO ] waiting for viable init node
> [INFO ] configuring bootstrap node(s) custom-9bdc82db3aed: waiting for agent to check in and apply initial plan
> [INFO ] configuring bootstrap node(s) custom-9bdc82db3aed: error applying plan -- check rancher-system-agent.service logs on node for more information, waiting for agent to check in and apply initial plan
> ```
> 
> Reading through the thread here it seems there isn't in a consensus on what can cause this problem directly other than "network issues". I was able to recreate the issue in my lab reusing a node from a previous cluster and then using a node with a clean install the node joins successfully, not sure what the common denominator is.


Follow [Certification](https://github.com/rancher/rancher/issues/41650#issuecomment-2075384843) did solve the problem for me. Here is my path to it.
![image](https://github.com/user-attachments/assets/be13178a-2fc5-4c5f-ae71-140d8c27d487)

**Comment by Oats87** [comment]
We really would need to implement some type of distro-specific pre-flight check to ensure a node can be provisioned.

Ref: 
https://github.com/rancher/rke2/issues/863 
https://github.com/k3s-io/k3s/issues/3167

It has been talked about on the distribution side before, but looks like nothing ever came of it.

The scope of provisioning ends when we pass the distro CLI and install script threshold, so log analysis and such of the distro does not fall to provisioning unless that is bubbled up through some type of interface. 

In the mean time, we can add additional hints to the error messages letting users know to check the logs for `rke2-server`, `rke2-agent`, `k3s`, and `k3s-agent` (dependent upon node role).

**Comment by woody41** [comment]
> This

Helped, but as someone mentioned, new image of the server didn't help me even though I added it without any issues for the first time! However, I got this bug only when I use custom domain. DNS was not being resolved from /etc/hosts and I had to manually edit `cattle-cluster-agent` pod and I added custom IP with domain. I directly edited
`kubectl edit deployment cattle-cluster-agent -n cattle-system` and added to `spec.template.spec.containers`

```
hostAliases:
    - ip: "1.2.3.4"  # IP of rancher server which generated joining command
      hostnames:
        - "my.domain" # And the domain
```

after this the node was added almost instantly. 🎉
